### PR TITLE
Connect API tokens to providers

### PR DESCRIPTION
**Pull request by tijmenb**
### Context

Vendors are only allowed to see and change applications that belong to the provider they have a key for.

### Changes proposed in this pull request

- Tokens are created for specific providers
- The API only exposes applications for the authenticating provider
- The API only allows updates to applications for the authenticating provider
- Test data generation is scoped to the authenticating provider 

### Guidance to review

- Does this make sense?
- Have we implemented enough authorization on the API with this change? 

I'll tidy up the commits after some initial review.

### Link to Trello card

https://trello.com/c/2vW0pk2z/1100-connect-api-tokens-to-providers

### CODE REVIEW

File: spec/support/vendor_api/vendor_api_spec_helpers.rb [hunk 1]
```diff
@@ -17,10 +17,14 @@ def post_api_request(url, options = {})
   end
 
   def auth_headers
-    unhashed_token = VendorApiToken.create_with_random_token!
+    unhashed_token = VendorApiToken.create_with_random_token!(provider: currently_authenticated_provider)
     { headers: { 'Authorization' => "Bearer #{unhashed_token}" } }
   end
 
+  def currently_authenticated_provider
```
Comment (duncanjbrown) [inline spec/support/vendor_api/vendor_api_spec_helpers.rb]: Elegant!

File: spec/requests/vendor_api/get_multiple_applications_spec.rb [hunk 2]
```diff
@@ -3,66 +3,51 @@
 RSpec.describe 'Vendor API - GET /api/v1/applications', type: :request do
   include VendorApiSpecHelpers
 
-  it 'returns applications scoped by `provider`' do
-    ucl         = create(:provider, code: 'UCL')
+  it 'returns applications scoped by the authenticating provider' do
     strathclyde = create(:provider, code: 'STR')
```
Comment (duncanjbrown) [inline spec/requests/vendor_api/get_multiple_applications_spec.rb]: There's a bit of dissonance giving the provider variable a name (and giving the provider itself a `code`) now that there's only one of them. Maybe leave the `provider` argument off the second call to `create_list`? (Which could probably just be a `create`, come to think of it)

File: app/controllers/vendor_api/vendor_api_controller.rb [hunk 3]
```diff
@@ -10,6 +10,11 @@ class VendorApiController < ActionController::API
 
   private
 
+    def provider_visible_applications
```
Comment (duncanjbrown) [inline app/controllers/vendor_api/vendor_api_controller.rb]: We already have `ApplicationChoice.for_provider` — could we push this scope onto the model and build on that? Others can benefit from it there too

File: app/views/support_interface/api_tokens/index.html.erb [hunk 4]
```diff
@@ -1,22 +1,25 @@
 <% content_for :title, 'API tokens' %>
 <h1 class='govuk-heading-xl'>API tokens</h1>
 
-<%= form_with url: support_interface_api_tokens_path, builder: GOVUKDesignSystemFormBuilder::FormBuilder do |f| %>
+<%= form_with model: VendorApiToken.new, url: support_interface_api_tokens_path, builder: GOVUKDesignSystemFormBuilder::FormBuilder do |f| %>
+  <%= f.govuk_collection_select :provider_id, Provider.all, :id, :name, label: { text: 'Provider' } %>
```
Comment (duncanjbrown) [inline app/views/support_interface/api_tokens/index.html.erb]: Bit squashed — could this control be bigger?

File: spec/requests/vendor_api/get_multiple_applications_spec.rb [hunk 5]
```diff
@@ -3,66 +3,51 @@
 RSpec.describe 'Vendor API - GET /api/v1/applications', type: :request do
   include VendorApiSpecHelpers
 
-  it 'returns applications scoped by `provider`' do
-    ucl         = create(:provider, code: 'UCL')
+  it 'returns applications scoped by the authenticating provider' do
```
Comment (duncanjbrown) [inline spec/requests/vendor_api/get_multiple_applications_spec.rb]: `to` the `authenticated`?

File: spec/requests/vendor_api/get_multiple_applications_spec.rb (hunk 2 above)
Comment (tijmenb) [reply spec/requests/vendor_api/get_multiple_applications_spec.rb]: Done

File: app/views/support_interface/api_tokens/index.html.erb (hunk 4 above)
Comment (tijmenb) [reply app/views/support_interface/api_tokens/index.html.erb]: I've added the least styling possible intentionally, so we can more easily update it later. Is that okay?

**Pull request merged**